mava/wrappers/gigastep.py

```diff
@@ -0,0 +1,332 @@
+        )
+        step_count = specs.BoundedArray(
+            (self.num_agents,), jnp.int32, 0, self._env.max_episode_length, "step_count"
+        )
+        if self.has_global_state:
+            number_of_agents_in_obs = (
+                self._env.n_agents if self.get_adversary_obs else self._env.n_agents_team1
+            )  # either the total num of agents or only team1 depending on the flag
+            adversary_action_sizes = (
+                self._env.n_agents_team2 if self.get_adversary_actions else 0
+            )  # consider the added space of the adversary actions
+            global_state = specs.BoundedArray(
+                (
+                    self.num_agents,
+                    self._env.observation_space.shape[0] * number_of_agents_in_obs
+                    + adversary_action_sizes,
+                ),
+                jnp.int32,
+                0,
+                255,
+                "global_state",
+            )
+            return specs.Spec(
+                ObservationGlobalState,
+                "ObservationSpec",
+                agents_view=agents_view,
+                action_mask=action_mask,
+                global_state=global_state,
+                step_count=step_count,
+            )
+
+        return specs.Spec(
+            Observation,
+            "ObservationSpec",
+            agents_view=agents_view,
+            action_mask=action_mask,
+            step_count=step_count,
+        )
+
+    def action_spec(self) -> specs.Spec:
+        return specs.MultiDiscreteArray(num_values=jnp.full(self.num_agents, self.num_actions))
+
+    def reward_spec(self) -> specs.Array:
+        return specs.Array(shape=(self.num_agents,), dtype=float, name="reward")
+
+    def discount_spec(self) -> specs.BoundedArray:
+        return specs.BoundedArray(
+            shape=(self.num_agents,), dtype=float, minimum=0.0, maximum=1.0, name="discount"
+        )
+
+    def _split_obs_and_state(
+        self, obs: Array, state: Tuple[Dict, Dict]
+    ) -> Tuple[Array, Tuple[Dict, Dict], Array, Tuple[Dict, Dict]]:
+        """
+        Separates the observations and state for both teams.
+
+        Args:
+            obs (Array): The observations of all agents.
+            state (Tuple[Dict, Dict]): The state of all agents.
+
+        Returns:
+            Tuple[Array, Tuple[Dict, Dict], Array, Tuple[Dict, Dict]]: Two tuples
+            representing observations and states for each team.
+        """
+        # The first n_agents_team1 elemnts in each array belong to team1
+        team1_obs, team2_obs = obs[: self.num_agents], obs[self.num_agents :]
+
+        # split each sub elemnt in the tuple
+        team1_state = jax.tree_util.tree_map(lambda x: x[: self.num_agents], state[0])
+        team2_state = jax.tree_util.tree_map(lambda x: x[self.num_agents :], state[0])
+
+        return team1_obs, (team1_state, state[1]), team2_obs, (team2_state, state[1])
```

Comment:
This is very confusing as to what `state[0]` and `state[1]` is, can you please split these out above the line e.g `state_<x>, state_<y> = state`